### DISCUSSION TITLE
Workaround to make `svelte-kit dev` and `svelte-kit build` work at the same time

### DISCUSSION THREAD

**Discussion by GauBen** [prisma/prisma#9027] [Q&A]
I found a workaround for an issue I have with Prisma, SvelteKit and Vite.

## Issue

Importing Prisma as ESM works well with `svelte-kit dev`:
```ts
import { PrismaClient } from '@prisma/client'
let prisma = new PrismaClient()
```

But breaks `svelte-kit build`:
```console
Named export 'PrismaClient' not found. The requested module '@prisma/client' is a CommonJS module, which may not support all module.exports as named exports.
CommonJS modules can always be imported via the default export, for example using:

import pkg from '@prisma/client';
const { PrismaClient } = pkg;
```

On the other hand, importing as CJS works well with `svelte-kit build`:
```ts
import Prisma from '@prisma/client'
let prisma = new Prisma.PrismaClient()
```

But breaks `svelte-kit dev`:
```console
TypeError: Cannot read property 'PrismaClient' of undefined
```

## Workaround

Create a file named `src/lib/prisma.ts`:
```ts
import type { PrismaClient } from '@prisma/client'
import Prisma from '@prisma/client'

export let prisma: PrismaClient

if (Prisma === undefined) {
  import('@prisma/client').then(({ PrismaClient }) => {
    prisma = new PrismaClient()
  })
} else {
  prisma = new Prisma.PrismaClient()
}
```

And import an already instantiated client with:
```ts
import { prisma } from '$lib/prisma'
// 🚀
```

Feel free to improve this workaround in the comments below. 👌

---

Related:
- [Support ES6 modules export #5030](https://github.com/prisma/prisma/issues/5030)
- [ feat: client esm support #4920](https://github.com/prisma/prisma/pull/4920)

**Comment by dslatkin** [comment]
@GauBen Did you figure out a way to instantiate the client only once across multiple hot module reloads of `prisma.ts`? I'm having issues defining a global variable to make it reusable while still using this shim. The Prisma docs recommend [doing it this way](https://www.prisma.io/docs/guides/performance-and-optimization/connection-management#prevent-hot-reloading-from-creating-new-instances-of-prismaclient).

**Reply by GauBen** [reply]
Does the following work for you?

```ts
import type { PrismaClient } from '@prisma/client'
import Prisma from '@prisma/client'

declare namespace globalThis {
  let prisma: PrismaClient
}

export let prisma: PrismaClient = globalThis.prisma

if (prisma === undefined) {
  if (Prisma === undefined) {
    import('@prisma/client').then(({ PrismaClient }) => {
      globalThis.prisma = prisma = new PrismaClient()
    })
  } else {
    globalThis.prisma = prisma = new Prisma.PrismaClient()
  }
}
```

**Reply by dslatkin** [reply]
I appreciate the help but I actually figured it out! I tried your solution just to see, but I got [a linting error](https://github.com/typescript-eslint/typescript-eslint/blob/v4.33.0/packages/eslint-plugin/docs/rules/no-namespace.md) on the `declare namespace` and runtime errors with the client. Not sure if that had to do with the method or the fact I'm running Node v16 which ended up being the main issue that was in my way.

What ended up working for me was this:

```ts
import pkg, { PrismaClient } from '@prisma/client';
import { dev } from '$app/env';

declare global {
    var _prisma: PrismaClient; // eslint-disable-line
}

let prisma;
if (dev) {
    if (!global._prisma) {
        global._prisma = new PrismaClient();
    }
    prisma = global._prisma;
} else {
    const { PrismaClient: PrismaClientProd } = pkg;
    prisma = new PrismaClientProd();
}

export default prisma as PrismaClient; // type assertion for shim
```

The line with `// eslint-disable-line` specifically is declared with `var` which I found out was necessary from [this Stack Overflow answer](https://stackoverflow.com/a/68328575) for Node v16 specifically. The ESM shim I used that worked was from [this GitHub comment](https://github.com/prisma/prisma/pull/4920#issuecomment-927334909).

**Comment by tnzk** [comment]
@GauBen I'm using a similar workaround but recently after bumped to `1.0.0-next.314` and switched to pnpm, I've been no longer able to use enum values correctly. Did you encountered similar problems?

**Reply by GauBen** [reply]
I never encountered your issue, you might want to open an issue to either svelte-kit or prisma.

**Reply by tnzk** [reply]
Thanks. Just curious, do you use enum types in your project?

**Reply by GauBen** [reply]
Nope, no enums so far

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
